src/main/java/io/pravega/perf/PerfStats.java

```diff
@@ -94,138 +182,197 @@ public void print(Instant time) {
+        private int[] getPercentiles() {
+            int[] percentileIds = new int[percentiles.length];
+            int[] values = new int[percentileIds.length];
+            int index = 0;
 
-        if (this.count % this.sampling == 0) {
-            this.latencies[index] = latency;
-            this.index++;
+            for (int i = 0; i < percentiles.length; i++) {
+                percentileIds[i] = (int) (count * percentiles[i]);
+            }
+
+            for (LatencyRange lr : latencyRanges) {
+                while ((index < percentileIds.length) &&
+                        (lr.start <= percentileIds[index]) && (percentileIds[index] <= lr.end)) {
+                    values[index++] = lr.latency;
+                }
+            }
+            return values;
         }
 
-        if (this.printer != null) {
-            final long nanotime = startTime.getEpochSecond() * NANOSEC_PER_SEC + startTime.getNano();
-            printer.printRecord(count, event, bytes, nanotime, latency);
+        public void record(int bytes, int latency) {
+            totalBytes += bytes;
+            latencies[latency]++;
         }
-    }
 
-    private static long[] percentiles(long[] latencies, int count, double... percentiles) {
-        int size = Math.min(count, latencies.length);
-        Arrays.sort(latencies, 0, size);
-        long[] values = new long[percentiles.length];
-        for (int i = 0; i < percentiles.length; i++) {
-            int index = (int) (percentiles[i] * size);
-            values[i] = latencies[index];
+        public void record(long start, int bytes, int latency) {
+            this.record(bytes, latency);
+        }
+
+        public void printTotal(long endTime) {
+            countLatencies();
+            final double elapsed = (endTime - startTime) / 1000.0;
+            final double recsPerSec = count / elapsed;
+            final double mbPerSec = (this.totalBytes / (1024.0 * 1024.0)) / elapsed;
+            int[] percs = getPercentiles();
+
+            System.out.printf(
+                    "%d records %s, %.3f records/sec, %d bytes record size, %.2f MB/sec, %.1f ms avg latency, %.1f ms max latency" +
+                            ", %d ms 50th, %d ms 75th, %d ms 95th, %d ms 99th, %d ms 99.9th\n",
+                    count, action, recsPerSec, messageSize, mbPerSec, totalLatency / ((double) count), (double) maxLatency,
+                    percs[0], percs[1], percs[2], percs[3], percs[4]);
         }
-        return values;
     }
 
-    /**
-     * print the performance statistics of current time window.
-     */
-    public synchronized void print() throws IOException {
-
-        final Instant time = Instant.now();
-        if (window.windowTimeMS(time) >= windowInterval) {
-            window.print(time);
-            this.window = new timeWindow();
-            if (printer != null) {
-                printer.flush();
+    @NotThreadSafe
+    static private class CSVLatencyWriter extends LatencyWriter {
```

Comment:
How do you guarantee that this class is used in a thread-safe manner?

Reply:
CSVLatencyWriter  is used as local variable inside the while call method the queue processor